feat: new cli commands `accounts:notes` and `accounts:transactions`

wd021 opened the pull request:
Proposing to add an `accounts:transactions` command to display transaction notes for an account. I've been debugging some account balance issues lately and found this to be quite helpful. Here's what it looks like

<img width="936" alt="Screen Shot 2022-02-28 at 2 04 16 PM" src="https://user-images.githubusercontent.com/2752586/155996526-74dd8a36-c884-4049-bf1c-964997b428b3.png">

If we add this, I'll open a PR to update docs on the website repo.

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
[x] No
```

dguenther commented:
Going to assign this to @NullSoldier because I think he wanted to give it a review next week before it merges. This looks useful!

QustomQure commented:
I think that this command must be able to work with both accountName and account address for better usability

wd021 commented:
for consistency, I'm using the same format as `accounts:balance`, which is accountName. We can update to take both, although it is one quick step to get names from `accounts:list`

Allmaaz commented:
Very good project

wd021 commented:
@dguenther @NullSoldier still interested in merging this? 

If so, i can update the code for latest version.. lmk!

dguenther commented:
I'd still like to merge this, I think it's definitely useful especially in the latest testnet phase. Nobody on our side has been directly involved in reviewing + merging it, so sorry for the delay! I'd like to get one other person than myself to take a look at it, then it should be good to go, if you don't mind updating the code to fix conflicts.

rohanjadvani commented on ironfish/src/account/accounts.ts:
```diff
@@ -596,6 +596,52 @@ export class Accounts {
+            memo: decryptedNote.memo().replace(/\x00/g, ''),
+          })
+
+          continue
```
Is this `continue` needed?

wd021 replied:
nope, removed.

xorinox commented on ironfish-cli/src/commands/accounts/transactions.test.ts:
```diff
@@ -0,0 +1,63 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { expect as expectCli, test } from '@oclif/test'
+import { GetTransactionsResponse } from 'ironfish'
```
This import didn't work for me, I had to change the path to  `../../../../ironfish/src/rpc/routes/transactions`

wd021 replied:
yup, this needs updating for the latest version. I just pushed an update.

mat-if commented:
Hey @wd021, I'm starting to dig into this PR, sorry for the delay. I wanted to clarify which direction to look at this from.

What's the intended goal of this CLI command? I think there's a few different things technically happening here so I wanted to make sure we're on the same page on expectations of what questions this command should answer.

For example, are you expecting this to show a complete list of all transactions this account has participated in? A complete list of notes, spent and unspent? A list of transactions containing unspent notes? etc.

Looking at this right now, it feels a little bit awkward in that it is kind of a view into transactions, and also kind of a view into notes, and I think we can improve that together

wd021 commented:
@mat-if no worries. 

The motivation for adding this command was to help debug account balances & faucet errors during phase 1. I wanted an easy way to output the spent notes associated with a given account.

I agree that the wording is probably unclear. Would also be interested in expanding the scope of this into filtering by txs/notes/unspent notes.

wd021 commented:
yea, 1 and 2 were exactly where my mind was at. let's have 2 separate methods, `accounts:transactions` and `accounts:notes`. And also agree that 3 could be useful too!

i don't mind taking a crack at all 3 in this PR, will have some time later in the week 👨‍💻

wd021 commented:
@mat-if PR updated. here's the latest👇 

`accounts:notes` - note level breakdown for given account (optional `-a` flag)

![Screenshot 2022-06-06 at 11 08 36](https://user-images.githubusercontent.com/2752586/172141378-4c29595a-c2c7-471d-8f29-2dc9dd217150.png)

`accounts:transactions` - transaction breakdown for given account or txhash (optional `-a` / `-t` flags)

![Screenshot 2022-06-06 at 11 08 56](https://user-images.githubusercontent.com/2752586/172141499-097a4105-90eb-4b39-a0f9-1e823ceb7bf6.png)

![Screenshot 2022-06-06 at 11 09 33](https://user-images.githubusercontent.com/2752586/172141504-ad9f9ed5-34fc-44df-90d3-73e975786cdf.png)

mat-if commented on ironfish-cli/src/commands/accounts/transactions.ts:
```diff
@@ -0,0 +1,113 @@
+
+    if (transactionInfo !== null) {
+      this.log(
+        `Transaction: ${transactionHash}, Status: ${transactionInfo.status}, Miner Fee: ${
```
What do you think about making each of these on a new line? I think it makes it a little easier to parse the info.
```
Transaction: fdaf29274587c8f58ac6214f42642fd610a222cea8e5e39ff4e11d58d379c194
Status: completed
Miner Fee: x
Fee ($ORE): 1
Spends: 1

```

wd021 replied:
update to this

mat-if commented on ironfish/src/account/accounts.ts:
```diff
@@ -921,6 +961,125 @@ export class Accounts {
+          }
+        }
+
+        if (transactionCreator) {
```
Same thing here about the `transactionCreator` thing, not able to look up a transaction that a note references if I only received in that transaction

![image](https://user-images.githubusercontent.com/97762857/172954975-41fbc895-61f8-4bfb-a87e-5f4502a4edc2.png)

wd021 replied:
yup, removed this.

mat-if reviewed (changes requested):
Thanks for doing this! A few comments, but overall this looks pretty good.

wd021 commented on ironfish/src/account/accounts.ts:
```diff
@@ -921,6 +961,125 @@ export class Accounts {
+    let transactionInfo = null
+    const transactionNotes = []
+
+    for (const transactionMapValue of this.transactionMap.values()) {
```
oh right, good catch.

wd021 commented on ironfish/src/account/accounts.ts:
```diff
@@ -603,6 +603,46 @@ export class Accounts {
     this.scan = null
   }
 
+  getNotesFor(account: Account): {
```
makes sense 👌

wd021 commented on ironfish/src/account/accounts.ts:
```diff
@@ -921,6 +961,125 @@ export class Accounts {
     await this.scanTransactions()
   }
 
+  getTransactionsFor(account: Account): {
```
makes sense 👌

wd021 commented on ironfish/src/account/accounts.ts:
```diff
@@ -921,6 +961,125 @@ export class Accounts {
+        }
+      }
+
+      if (transactionCreator) {
```
right. I updated this method to include notes where one is the receiver of iron in a transaction. but i also think its valuable to be able to detect the transactions created by the account, so there's now an extra `creator` column for the table.

wd021 commented:
@mat-if 
- went through review changes and updated
- added extra `creator` column for account:transactions
- merge latest staging changes to resolve merge conflict

mat-if reviewed (approved):
Thanks!
